### PR TITLE
Add Better Egg (v2) scoring model

### LINKED ISSUE #30: Implement Good Egg v2 scoring model
## Overview

Implement a v2 scoring model based on findings from the validation study (`experiments/ge-validation` branch, PR #27). The v2 model simplifies the trust graph (removing 3 ineffective dimensions), adds external features (temporal merge rate, account age), and combines them via a trained logistic regression. Text dissimilarity is supported but off by default.

**Backwards compatibility:** v2 is opt-in. The default behavior remains the v1 scoring model. Users enable v2 via configuration (`scoring_model: v2`). All existing config keys continue to work.

---

### 1. Simplify the trust graph (drop 3 dimensions)

Remove three graph dimensions that have zero measurable impact (all adj. p > 0.96):

| Remove | Current config key | Current default |
|--------|-------------------|-----------------|
| Self-contribution penalty | `edge_weights.self_contribution_penalty` | 0.3 |
| Diversity/volume scaling | `graph_scoring.diversity_scale` / `volume_scale` | 0.5 / 0.3 |
| Language normalization | `language_normalization.multipliers` + `.default` | 28-entry table + 3.0 |

**Implementation:** In v2 mode, hard-set `self_contribution_penalty=1.0` (no penalty), `diversity_scale=0.0`, `volume_scale=0.0`, and skip the language normalization lookup. The 6 retained parameters are: `half_life_days`, `max_age_days`, `archived_penalty`, `fork_penalty`, `MAX_PRS_PER_REPO`, `reverse_edge_ratio`.

Repo quality (stars, archived, fork) and language match (personalization weighting) are **retained** — both showed small but significant effects.

### 2. Add temporal merge rate feature

Compute the author's merge rate using only PRs merged/closed *before* the PR being scored:

```
merged_before_T = count(author merged PRs where merged_at < T)
closed_before_T = count(author closed PRs where closed_at < T)
temporal_merge_rate = merged_before_T / (merged_before_T + closed_before_T)
```

**Data requirement:** Fetch closed PR timestamps via `pullRequests(states: CLOSED)` in the existing GraphQL query. Add `closed_prs` to `UserProfile` or extend the contribution data model.

**Edge case:** If denominator is 0 (no prior PRs), treat merge rate as missing/null. The combined model should handle missing features gracefully (e.g., impute population mean or use a reduced model).

### 3. Add account age feature

```python
log_account_age = math.log(account_age_days + 1)
```

Already available from `UserProfile.created_at`. Just needs to be surfaced as a scoring input in v2 mode.

### ~~4. Add text dissimilarity feature (off by default)~~

```python
text_dissimilarity = 1.0 - tfidf_cosine_similarity(pr_text, repo_readme)
```

~~**Off by default.** Enable via config:~~

```yaml
v2:
  text_dissimilarity:
    enabled: false          # default
    method: tfidf           # tfidf | jaccard
    max_pr_text_chars: 2000
    max_readme_chars: 4000
```

~~When enabled, requires PR body text and repo README as inputs. Use TF-IDF cosine similarity as the default method. Jaccard as a simpler alternative.~~

~~**Rationale for off-by-default:** The inverted direction (higher similarity → lower merge probability) is counterintuitive and the feature could be gamed by submitting intentionally divergent PR descriptions. It improves AUC in controlled study conditions but is not obviously justified as a production decision metric.~~

**Update:** Not implementing for now until this phenomenon can be understood as something other than noise.

### 5. Combined model (Layer 3)

Logistic regression combining graph score with external features:

```python
P(merge) = sigmoid(w0 + w1 * graph_score + w2 * merge_rate + w3 * log_age [+ w4 * dissimilarity])
```

- `penalty=None` (unregularized — 3-4 features with thousands of observations)
- Ship default weights trained on the validation study dataset (5,417 PRs, 49 repos)
- Store weights in a config-loadable format (YAML or JSON)
- The number of active weights depends on which features are enabled

### 6. Configuration

```yaml
scoring_model: v1          # v1 (default) | v2

# v2-specific settings (ignored when scoring_model=v1)
v2:
  graph:
    half_life_days: 180
    max_age_days: 730
    archived_penalty: 0.5
    fork_penalty: 0.3
  features:
    temporal_merge_rate: true    # requires closed PR data
    account_age: true
  text_dissimilarity:
    enabled: false
    method: tfidf
  combined_model:
    weights_file: null          # path to custom weights; null = use built-in defaults
```

### 7. Scoring API changes

`TrustScorer.score()` in v2 mode should:
1. Build the simplified graph (no self-penalty, no diversity/volume, no language norm)
2. Compute graph score via existing graph scoring pipeline
3. Compute external features (merge rate, age, optionally dissimilarity)
4. Run combined model to produce `P(merge)`
5. Classify into trust levels using existing thresholds
6. Return `TrustScore` with the combined probability as the primary score

The `TrustScore` model should be extended to include the component scores (graph score, merge rate, age, dissimilarity) for transparency.

---

## Out of scope

- Running stale threshold lookback (future optimization)
- Per-repo model fine-tuning (future)
- Newcomer-specific model (future)
- Gaming resistance testing (future)
- Temporal merge rate trends (future)
- Repository-specific calibration (future)

## References

- Validation study: PR #27, branch `experiments/ge-validation`
- Full specification: `experiments/validation/results/BETTER_EGG.md`
- Statistical results: `experiments/validation/results/statistical_tests.json`
- Study summary: `experiments/validation/results/SUMMARY.md`

### PR DESCRIPTION
**Pull request by jeffreyksmithjr**
## Summary

- Add opt-in v2 scoring model ("Better Egg") based on the validation study (#27): simplified graph + logistic regression combining graph score, merge rate, and account age
- Simplify graph builder by removing 3 zero-signal dimensions (self-contribution penalty, language normalization in repo quality, diversity/volume adjustment); language match personalization is retained per validation findings
- Add closed PR count to GraphQL query at zero API cost for merge rate computation
- Add `scoring_model` config (`v1` default, `v2` opt-in) with `--scoring-model` CLI option, `scoring-model` action input/output, MCP parameter, and `GOOD_EGG_SCORING_MODEL` env var
- Add "Better Egg" branding and component score breakdown to all output formatters
- Text dissimilarity intentionally excluded — inverted signal not yet well enough understood
- 56 new tests (269 total, 93% coverage), full documentation, examples, and CHANGELOG

## Test plan

- [x] `uv run pytest --cov=good_egg -v` — 269 tests pass, 93% coverage
- [x] `uv run ruff check src/ tests/ scripts/` — all checks pass
- [x] `uv run mypy src/good_egg/` — no issues found
- [x] Manual smoke test: `good-egg score jeffreyksmithjr --repo 2ndSetAI/good-egg --scoring-model v2 --verbose` — Better Egg: HIGH (79%)
- [x] Verify v1 output is unchanged: `--scoring-model v1` — Good Egg: HIGH (85%), empty component_scores, scoring_model="v1"
- [x] Review PR comment rendering with v2 "Better Egg" branding — CI `score` check passed using `@better-egg` ref with `scoring-model: v2`

## Review findings addressed

Red team review identified 3 critical, 6 high, 7 medium, 4 low, and 6 informational issues. All were addressed in f6946d6 except text dissimilarity (intentionally excluded). Key fixes:

- Removed ghost `combined_score` from all docs (C1)
- Fixed config field names and weight values in docs to match code (C2, C3)
- Restored `same_language_weight` in v2 per issue spec (H2)
- Explained negative merge_rate_weight in methodology docs (H4)
- Replaced unsafe Pydantic mutation with `model_copy()` (H6)
- Added 3 tests for edge cases: all features disabled, negative weight, opposing signals (M5-M7)
- Standardized Good Egg (v1) / Better Egg (v2) naming across all docs

Closes #30

### CODE REVIEW

File: docs/methodology.md [hunk 1]
```diff
@@ -145,3 +145,156 @@ GitHub rate limits bound how much data can be fetched per user. Good Egg is desi
...
+is used as the final normalized score. The trained weights are:
+
+| Coefficient | Value |
+|-------------|-------|
+| `intercept` | -0.8094 |
+| `graph_score_weight` | 1.9138 |
+| `merge_rate_weight` | -0.7783 |
+| `account_age_weight` | 0.1493 |
+
+These were trained on the validation dataset (see below).
```
Comment (rlronan) [inline docs/methodology.md]: Might be clearer to say validation study data
